staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go

```diff
@@ -279,102 +305,277 @@ func (o *ExposeServiceOptions) RunExpose(cmd *cobra.Command, args []string) erro
 					return cmdutil.UsageErrorf(cmd, "couldn't find port via --port flag or introspection")
 				}
 			case 1:
-				params["port"] = ports[0]
+				o.Port = ports[0]
 			default:
-				params["ports"] = strings.Join(ports, ",")
+				o.Ports = strings.Join(ports, ",")
 			}
 		}
 
 		// Always try to derive protocols from the exposed object, may use
 		// different protocols for different ports.
-		if _, found := params["protocol"]; found {
-			protocolsMap, err := o.ProtocolsForObject(info.Object)
-			if err != nil {
-				return cmdutil.UsageErrorf(cmd, "couldn't find protocol via introspection: %v", err)
-			}
-			if protocols := generate.MakeProtocols(protocolsMap); !generate.IsZero(protocols) {
-				params["protocols"] = protocols
-			}
+		protocolsMap, err := o.ProtocolsForObject(info.Object)
+		if err != nil {
+			return cmdutil.UsageErrorf(cmd, "couldn't find protocol via introspection: %v", err)
+		}
+		if protocols := makeProtocols(protocolsMap); protocols != "" {
+			o.Protocols = protocols
 		}
 
-		if generate.IsZero(params["labels"]) {
+		if len(o.Labels) == 0 {
 			labels, err := meta.NewAccessor().Labels(info.Object)
 			if err != nil {
 				return err
 			}
-			params["labels"] = polymorphichelpers.MakeLabels(labels)
-		}
-		if err = generate.ValidateParams(names, params); err != nil {
-			return err
-		}
-		// Check for invalid flags used against the present generator.
-		if err := generate.EnsureFlagsValid(cmd, generators, generatorName); err != nil {
-			return err
+			o.Labels = polymorphichelpers.MakeLabels(labels)
 		}
 
-		// Generate new object
-		object, err := generator.Generate(params)
+		// Generate new service
+		service, err := o.createService()
 		if err != nil {
 			return err
 		}
 
 		if inline := cmdutil.GetFlagString(cmd, "overrides"); len(inline) > 0 {
 			codec := runtime.NewCodec(scheme.DefaultJSONEncoder(), scheme.Codecs.UniversalDecoder(scheme.Scheme.PrioritizedVersionsAllGroups()...))
-			object, err = cmdutil.Merge(codec, object, inline)
+			service, err := cmdutil.Merge(codec, service, inline)
+			if err != nil {
+				return err
+			}
+			err = o.checkService(service, cmd)
 			if err != nil {
 				return err
 			}
 		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
You forgot to add a `o.checkService` after, and using `var service *corev1.Service` will increase the readability 
```suggestion
var service *corev1.Service
		// Generate new service
		service, err = o.createService()
		if err != nil {
			return err
		}

		if inline := cmdutil.GetFlagString(cmd, "overrides"); len(inline) > 0 {
			codec := runtime.NewCodec(scheme.DefaultJSONEncoder(), scheme.Codecs.UniversalDecoder(scheme.Scheme.PrioritizedVersionsAllGroups()...))
			service, err := cmdutil.Merge(codec, service, inline)
			if err != nil {
				return err
			}
			err = o.checkService(service, cmd)
			if err != nil {
				return err
			}
		}

		err = o.checkService(service, cmd)
		if err != nil {
			return err
		}
		return nil
```